source/common/buffer/buffer_impl.h

```diff
@@ -10,6 +10,44 @@
 namespace Envoy {
 namespace Buffer {
 
+/**
+ * An implementation of BufferFragment where a releasor callback is called when the data is
+ * no longer needed.
+ */
+class BufferFragmentImpl : public BufferFragment {
+public:
+  /**
+   * Creates a new wrapper around the externally owned <data> of size <size>.
+   * The caller must ensure <data> is valid until releasor() is called, or for the lifetime of the
+   * fragment.
+   * @param data external data to reference
+   * @param size size of data
+   * @param releasor a callback function to be called when data is no longer needed.
+   */
+  BufferFragmentImpl(const void* data, size_t size,
+                     std::function<void(const void*, size_t)> releasor)
+      : data_(data), size_(size), releasor_(releasor) {}
+
+  ~BufferFragmentImpl() override {}
+
+  const void* data() override { return data_; }
+  size_t size() override { return size_; }
+
+  void done() override {
+    if (releasor_) {
+      releasor_(data_, size_);
+    }
+  }
+
+  BufferFragmentImpl(const BufferFragmentImpl&) = delete;
+  BufferFragmentImpl& operator=(const BufferFragmentImpl&) = delete;
+
+private:
+  const void* data_;
```

Comment:
These should all be `const`, e.g. `const void* const data`.

Reply:
You mean just the pointer? or the members too?

Reply:
The others as well; they are only set in the constructor initializor list.